include/seastar/core/memory.hh

```diff
@@ -240,38 +240,38 @@ class statistics {
             uint64_t total_memory, uint64_t free_memory, uint64_t reclaims, uint64_t large_allocs,
-            uint64_t foreign_mallocs, uint64_t foreign_frees, uint64_t foreign_cross_frees)
+            uint64_t foreign_mallocs, uint64_t foreign_frees, uint64_t foreign_cross_frees) noexcept
         : _mallocs(mallocs), _frees(frees), _cross_cpu_frees(cross_cpu_frees)
         , _total_memory(total_memory), _free_memory(free_memory), _reclaims(reclaims), _large_allocs(large_allocs)
         , _foreign_mallocs(foreign_mallocs), _foreign_frees(foreign_frees)
         , _foreign_cross_frees(foreign_cross_frees) {}
 public:
     /// Total number of memory allocations calls since the system was started.
-    uint64_t mallocs() const { return _mallocs; }
+    uint64_t mallocs() const noexcept { return _mallocs; }
     /// Total number of memory deallocations calls since the system was started.
-    uint64_t frees() const { return _frees; }
+    uint64_t frees() const noexcept { return _frees; }
     /// Total number of memory deallocations that occured on a different lcore
     /// than the one on which they were allocated.
-    uint64_t cross_cpu_frees() const { return _cross_cpu_frees; }
+    uint64_t cross_cpu_frees() const noexcept { return _cross_cpu_frees; }
     /// Total number of objects which were allocated but not freed.
-    size_t live_objects() const { return mallocs() - frees(); }
+    size_t live_objects() const noexcept { return mallocs() - frees(); }
     /// Total free memory (in bytes)
-    size_t free_memory() const { return _free_memory; }
+    size_t free_memory() const noexcept { return _free_memory; }
     /// Total allocated memory (in bytes)
-    size_t allocated_memory() const { return _total_memory - _free_memory; }
+    size_t allocated_memory() const noexcept { return _total_memory - _free_memory; }
     /// Total memory (in bytes)
-    size_t total_memory() const { return _total_memory; }
+    size_t total_memory() const noexcept { return _total_memory; }
     /// Number of reclaims performed due to low memory
-    uint64_t reclaims() const { return _reclaims; }
+    uint64_t reclaims() const noexcept { return _reclaims; }
     /// Number of allocations which violated the large allocation threshold
-    uint64_t large_allocations() const { return _large_allocs; }
+    uint64_t large_allocations() const noexcept { return _large_allocs; }
     /// Number of foreign allocations
-    uint64_t foreign_mallocs() const { return _foreign_mallocs; }
+    uint64_t foreign_mallocs() const noexcept { return _foreign_mallocs; }
     /// Number of foreign frees
-    uint64_t foreign_frees() const { return _foreign_frees; }
+    uint64_t foreign_frees() const noexcept { return _foreign_frees; }
     /// Number of foreign frees on reactor threads
-    uint64_t foreign_cross_frees() const { return _foreign_cross_frees; }
-    friend statistics stats();
+    uint64_t foreign_cross_frees() const noexcept { return _foreign_cross_frees; }
```

Comment:
I see why one would want to mark methods that are part of the public API as noexcept, but what is the point in marking CU internal methods? IMO it only adds noise.

Reply:
It is a contract that any function that uses them can rely on and be made or simply marked noexcept.
If that changes at the low-level / private functions, anyone who changes them should make sure what breaks if the function throws and propagate the change up the stack.  Otherwise, noone would know or care if the function doesn't guarantee any exception semantics.

Reply:
This almost sounds like the now retired throw specifications, which were retired exactly because a change in a low-level function triggered excessive changes across the code-base, as all dependent methods had to also update their throw-specification. This is what excessive usage of noexcept will cause too (or alternatively, crashes).
One other thing to think about is that marking library methods as noexcept means we commit to not use exceptions in the future, meaning that error handling will be limited to logging or aborting. This is something that shouldn't be taken lightly.